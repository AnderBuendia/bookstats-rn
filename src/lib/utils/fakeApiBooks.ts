import { BookStatus } from 'enums/book-status.enum';
import type { Book } from '@Interfaces/domain/book.interface';

const date = new Date('2022');

export const HomeBooks: Book[] = [
  {
    id: '1',
    title: 'The Brothers Karamazov',
    author: 'Dostoievsky',
    status: BookStatus.COMPLETED,
    rating: 80,
    pages: 0,
    userId: 'user1',
    image: null,
    review: null,
    read_pages: [],
    createdAt: date,
    updatedAt: date,
  },
  {
    id: '2',
    title: 'Steppenwolf',
    author: 'Hesse',
    status: BookStatus.READING,
    rating: 40,
    pages: 20,
    userId: 'user2',
    image: null,
    review: null,
    read_pages: [],
    createdAt: date,
    updatedAt: date,
  },
  {
    id: '3',
    title: 'One hundred years of solitude',
    author: 'MÃ¡rquez',
    status: BookStatus.READY,
    rating: 20,
    pages: 417,
    userId: 'user3',
    image: null,
    review: null,
    read_pages: [],
    createdAt: date,
    updatedAt: date,
  },
  {
    id: '4',
    title: 'The Symposium',
    author: 'Plato',
    status: BookStatus.TO_READ,
    rating: 0,
    pages: 111,
    userId: 'user4',
    image: null,
    review: null,
    read_pages: [],
    createdAt: date,
    updatedAt: date,
  },
];
