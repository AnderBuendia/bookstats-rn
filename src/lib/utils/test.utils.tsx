import type { ReactElement } from 'react';
import { render } from '@testing-library/react-native';
import { AppStoreContext } from '@Lib/context/app-store.context';
import AuthStoreContext from 'lib/context/auth-store.context';
import { UIState } from '@Enums/config/ui-state.enum';
import type { ErrorState } from '@Interfaces/config/errors.interface';

export interface RenderOptions {
  providerProps?: {
    uiState?: UIState | null;
    errorState?: ErrorState;
    setUiState?: () => void;
    setErrorState?: () => void;
  };
  renderOptions?: {};
}

const customRender = (ui: ReactElement, options: RenderOptions = {}) => {
  const authProviderValues = {
    uiState: options.providerProps?.uiState,
    setUiState: options.providerProps?.setUiState,
  };

  const appProviderValues = {
    errorState: options.providerProps?.errorState,
    setErrorState: options.providerProps?.setErrorState,
  };

  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      <AuthStoreContext.Provider value={authProviderValues}>
        <AppStoreContext.Provider value={appProviderValues}>
          {children}
        </AppStoreContext.Provider>
      </AuthStoreContext.Provider>
    ),
    ...options.renderOptions,
  });
};

export * from '@testing-library/react-native';
// override render export
export { customRender as render };
