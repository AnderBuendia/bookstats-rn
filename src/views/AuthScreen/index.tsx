import { AntDesign } from '@expo/vector-icons';
import { StatusBar } from 'expo-status-bar';
import { FC, useEffect } from 'react';
import { Image, Platform, Pressable, StyleSheet } from 'react-native';
import { useAuthUserStorage } from '@Services/storage.service';
import { useErrorStorage } from '@Services/storage.service';
import Colors from '@Lib/constants/Colors';
import { Text, View } from '@Components/generic/Theme/Themed';
import LoginForm from '@Components/Forms/LoginForm';
import SignUpForm from '@Components/Forms/SignUpForm';
import ConfirmSignUpForm from '@Components/Forms/ConfirmSignUpForm';
import { UIState } from '@Enums/config/ui-state.enum';
import type { RootStackScreenProps } from '@Types/main.type';

export type AuthScreenProps = RootStackScreenProps<'Auth'>;

const AuthScreen: FC<AuthScreenProps> = ({ navigation }) => {
  const { uiState, setUiState, user } = useAuthUserStorage();
  const { errorState, setErrorState } = useErrorStorage();
  const sourceImage = require('../../../assets/images/book.png');

  useEffect(() => {
    const timeoutId = setTimeout(
      () => setErrorState({ hasError: false }),
      4000
    );

    return () => clearTimeout(timeoutId);
  }, [errorState]);

  const uiStateForm = (uiState: UIState) => {
    switch (uiState) {
      case UIState.SIGN_IN:
        return <LoginForm navigation={navigation} handleUiState={setUiState} />;
      case UIState.CONFIRM_SIGN_UP:
        return <ConfirmSignUpForm handleUiState={setUiState} />;
      case UIState.SIGN_UP:
        return <SignUpForm handleUiState={setUiState} />;
      default:
        return <LoginForm navigation={navigation} handleUiState={setUiState} />;
    }
  };

  return (
    <View style={styles.loginContainer}>
      <View style={styles.header}>
        <Pressable onPress={() => navigation.navigate('Home')}>
          <AntDesign
            name="arrowleft"
            size={24}
            color={Colors.light.background}
          />
        </Pressable>

        <Image style={styles.headerImage} source={sourceImage} />
        <Text style={styles.headerTitle}>Bookstats</Text>
      </View>

      <View style={styles.formContainer}>
        {errorState.hasError && (
          <Text style={styles.errorMessage_text}>{errorState.message}</Text>
        )}

        {uiState && uiStateForm(uiState)}
      </View>

      {/* Use a light status bar on iOS to account for the black space above the modal */}
      <StatusBar style={Platform.OS === 'ios' ? 'light' : 'auto'} />
    </View>
  );
};

const styles = StyleSheet.create({
  loginContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    padding: 20,
  },
  headerImage: {
    marginLeft: 14,
    width: 36,
    height: 36,
  },
  headerTitle: {
    marginLeft: 4,
    fontWeight: 'bold',
    fontSize: 18,
  },
  errorMessage_text: {
    position: 'absolute',
    top: 0,
    textAlign: 'center',
    width: 350,
    padding: 2,
    borderWidth: 2,
    borderColor: Colors.error_300.background,
    borderRadius: 6,
    color: Colors.error_100.text,
    backgroundColor: Colors.error_700.background,
  },
  formContainer: {
    flex: 2,
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default AuthScreen;
