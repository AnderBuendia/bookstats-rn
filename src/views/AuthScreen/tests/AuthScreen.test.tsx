import Amplify, { Auth } from 'aws-amplify';
import awsconfig from '../../../../aws-exports';
Amplify.configure(awsconfig);
import { Button, Text, TextInput, View } from 'react-native';
import { render, fireEvent, waitFor } from '@Lib/utils/test.utils';
import AuthScreen from '@Views/AuthScreen';
import { UIState } from '@Enums/config/ui-state.enum';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...props,
});

Auth.signIn = jest.fn().mockImplementation(() => {
  return true;
});

describe('Test Auth Screen', () => {
  let props: any;

  beforeEach(() => {
    props = createTestProps({});
  });

  it('Should show Login title text', async () => {
    const providerProps = {
      uiState: UIState.SIGN_IN,
      errorState: { hasError: false },
      setUiState: () => {},
    };

    render(<AuthScreen {...props} />, { providerProps });

    expect(/Login/i).toMatchSnapshot();
  });
});
