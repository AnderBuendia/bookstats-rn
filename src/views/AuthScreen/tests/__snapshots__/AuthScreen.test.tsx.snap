// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Auth Screen Should show Books title text 1`] = `/Login/i`;

exports[`Test Auth Screen Should show Login title text 1`] = `/Login/i`;
